### PR TITLE
Add inline assembly internals

### LINKED ISSUE #686: Document inline asm
@Amanieu has written some notes here: https://github.com/rust-lang/rust/pull/69171#issue-375572066

(Sorry, I know we are attempting to reduce this sort of post in this sprint)...

### PR DESCRIPTION
**Pull request by Amanieu**
This PR adds information about the internals of inline assembly in the compiler, including instructions to follow when adding `asm!` support for a new architecture.

Fixes #686

### CODE REVIEW

**Review by jyn514 — APPROVED** [review]
This is awesome! Thanks for writing it up :)

**Comment by camelid** [comment]
Is this waiting on another review, e.g. from somewhat knowledgeable about the `asm!` internals?

**Comment by camelid** [comment]
@joshtriplett could you review this? I seem to recall you're knowledgeable about inline asm. Or, if you can't review it, could you let me know who might be able to?

**Review by spastorino — APPROVED** [review]

**Comment by spastorino** [comment]
I think this is great enough to be merged, we can always iterate it and make it better if we will as with any other section of the guide. Merging ...

**Pull request merged**
